### ISSUE TITLE
Header file causes clangd to use excessive amount of memory

### ISSUE THREAD

**Issue by xenu** [clangd/clangd#2098]
Opening the attached header file ([perl.h.txt](https://github.com/user-attachments/files/16339549/perl.h.txt)) in an editor causes clangd's memory usage to rapidly grow. I killed the process when it reached >10GB (which took ~1 minute).

There is no `compile_commands.json` or `compile_flags.txt`.

**Logs**

[log.txt](https://github.com/user-attachments/files/16339993/log.txt)

**System information**

Output of `clangd --version`:

```
clangd version 18.1.8
Features: windows
Platform: x86_64-pc-windows-msvc
```

I can also reproduce it with clangd 14.0.6 on Linux, so I don't think it's an environment-specific issue.

Editor/LSP plugin:

It happens with both Emacs+Eglot and VSCode + clangd extension (latest versions).

Operating system:

Windows and Linux

**Comment by HighCommander4** [comment]
Could you try with the latest weekly snapshot at https://github.com/clangd/clangd/releases?

I landed a fix in clangd 19 (#1384) which should hopefully fix many issues of this sort.

**Comment by xenu** [comment]
Thank you! With the snapshot the memory usage is just 28 MB.
